src/track/beats.cpp

```diff
@@ -32,25 +541,474 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+
+        if (currentBpmMarker.beat_index() == m_beats.size()) {
+            markers |= Beat::Marker::BPM;
+        }
+        auto currentTimeSignatureMarker =
+                m_beatsProto.time_signature_markers().Get(
+                        timeSignatureMarkerIndex);
+        TimeSignature currentTimeSignature =
+                TimeSignature(currentTimeSignatureMarker.time_signature());
+        FrameDiff_t beatLength = getBeatLengthFrames(
+                bpmBeforeThisBeat, m_iSampleRate, currentTimeSignature);
+
+        if (barRelativeBeatIndex % currentTimeSignature.getBeatsPerBar() == 0) {
+            barIndex++;
+            if (m_beatsProto.first_downbeat_index() == m_beats.size()) {
+                markers |= Beat::Marker::TIME_SIGNATURE;
+            }
+            if (timeSignatureMarkerIndex <
+                            m_beatsProto.time_signature_markers_size() - 1 &&
+                    m_beatsProto.time_signature_markers()
+                                    .Get(timeSignatureMarkerIndex + 1)
+                                    .downbeat_index() == barIndex) {
+                timeSignatureMarkerIndex++;
+                currentTimeSignatureMarker =
+                        m_beatsProto.time_signature_markers().Get(
+                                timeSignatureMarkerIndex);
+                currentTimeSignature = TimeSignature(
+                        currentTimeSignatureMarker.time_signature());
+                markers |= Beat::Marker::TIME_SIGNATURE;
+            }
+        }
+
+        Beat::Type beatType =
+                barRelativeBeatIndex % currentTimeSignature.getBeatsPerBar() ==
+                        0
+                ? Beat::DOWNBEAT
+                : Beat::BEAT;
+
+        FramePos beatFramePosition = m_beats.empty()
+                ? FramePos(m_beatsProto.first_frame_position())
+                : (m_beats.last().getFramePosition() + beatLength);
+        addedBeat = Beat(beatFramePosition,
+                beatType,
+                currentTimeSignature,
+                Bpm(currentBpmMarker.bpm()),
+                m_beats.size(),
+                barIndex,
+                barRelativeBeatIndex,
+                markers);
+        barRelativeBeatIndex = (barRelativeBeatIndex + 1) %
+                currentTimeSignature.getBeatsPerBar();
+        if (addedBeat.getFramePosition() <= trackLastFrame) {
+            m_beats.append(addedBeat);
+        } else {
+            break;
+        }
+    }
+    updateBpm();
+}
+void BeatsInternal::clearMarkers() {
+    m_beatsProto.clear_first_frame_position();
+    m_beatsProto.clear_first_downbeat_index();
+    m_beatsProto.clear_bpm_markers();
+    m_beatsProto.clear_time_signature_markers();
+    m_beats.clear();
+}
+
+void BeatsInternal::setAsDownbeat(int beatIndex) {
+    auto beat = getBeatAtIndex(beatIndex);
+    m_beatsProto.set_first_downbeat_index(m_beatsProto.first_downbeat_index() +
```

Comment:
This seems problematic. If I understand correctly, shifting any downbeat shifts the downbeats for the entire track; there is no way to shift the downbeats for part of the track. I think we need to be able to shift downbeats between arbitrary frames or beat indices.

Reply:
On further thought, this may be okay. Having only this one index to shift downbeats ensures that all the downbeat markers are consistent with the time signatures.